criu/arch/x86/shstk.c

```diff
@@ -88,3 +91,133 @@ int arch_shstk_prepare(struct pstree_item *item, CoreEntry *core,
 
 	return 0;
 }
+
+int arch_shstk_unlock(struct pstree_item *item, CoreEntry *core, pid_t pid)
+{
+	unsigned long features;
+	int status;
+	int ret = -1;
+
+	/*
+	 * CRIU runs with no shadow stack and the task does not need one,
+	 * nothing to do.
+	 */
+	if (!kdat.has_shstk && !task_needs_shstk(item, core))
+		return 0;
+
+	futex_wait_until(&rsti(item)->shstk_enable, 1);
+
+	if (ptrace(PTRACE_SEIZE, pid, 0, 0)) {
+		pr_perror("Cannot attach to %d", pid);
+		goto futex_wake;
+	}
+
+	if (ptrace(PTRACE_INTERRUPT, pid, 0, 0)) {
+		pr_perror("Cannot interrupt the %d task", pid);
+		goto detach;
+	}
+
+	if (wait4(pid, &status, __WALL, NULL) != pid) {
+		pr_perror("waitpid(%d) failed", pid);
+		goto detach;
+	}
+
+	features = ARCH_SHSTK_SHSTK | ARCH_SHSTK_WRSS;
+	if (ptrace(PTRACE_ARCH_PRCTL, pid, features, ARCH_SHSTK_UNLOCK)) {
+		pr_perror("Cannot unlock CET for %d task", pid);
+		goto detach;
+	}
+
+detach:
+	if (ptrace(PTRACE_DETACH, pid, NULL, 0)) {
+		pr_perror("Unable to detach %d", pid);
+		goto futex_wake;
+	}
+
+	ret = 0;
+
+futex_wake:
+	futex_set_and_wake(&rsti(item)->shstk_unlock, 1);
+
+	return ret;
+}
+
+static void shstk_sync_unlock(struct pstree_item *item)
+{
+	/* notify parent that shadow stack is enabled ... */
+	futex_set_and_wake(&rsti(item)->shstk_enable, 1);
+
+	/* ... and wait until it unlocks its features with ptrace */
+	futex_wait_until(&rsti(item)->shstk_unlock, 1);
+}
+
+static void __arch_shstk_enable(struct pstree_item *item,
+				int (*func)(void *arg), void *arg)
+{
+	int ret;
+
+	shstk_sync_unlock(item);
+
+	/* return here would cause #CP, use exit() instead */
+	ret = func(arg);
+	exit(ret);
+}
+
+static int shstk_disable(struct pstree_item *item)
+{
+	shstk_sync_unlock(item);
+
+	/* disable shadow stack, implicitly clears ARCH_SHSTK_WRSS */
+	if (syscall(__NR_arch_prctl, ARCH_SHSTK_DISABLE, ARCH_SHSTK_SHSTK)) {
+		pr_perror("Failed to disable shadow stack");
+		return -1;
+	}
+
+	if (syscall(__NR_arch_prctl, ARCH_SHSTK_LOCK,
+		    ARCH_SHSTK_SHSTK | ARCH_SHSTK_WRSS)) {
+		pr_perror("Failed to lock shadow stack controls");
+		return -1;
+	}
+
+	return 0;
+}
+
+int arch_shstk_trampoline(struct pstree_item *item, CoreEntry *core,
+		      int (*func)(void *arg), void *arg)
+{
+	unsigned long features = ARCH_SHSTK_SHSTK;
+	int code = ARCH_SHSTK_ENABLE;
+
+	/*
+	 * If task does not need shadow stack but CRIU runs with shadow
+	 * stack enabled, we should disable it before continuing with
+	 * restore
+	 */
+	if (!task_needs_shstk(item, core)) {
+		if (kdat.has_shstk && shstk_disable(item))
+			return -1;
+		return func(arg);
```

Comment:
It might be useful for debugging to add a log message here, indicating that the shadow stack is disabled.